amo/blockchain/hibernate_test.go

```diff
@@ -186,3 +188,137 @@ func TestMissStat(t *testing.T) {
 	assert.Equal(t, int64(11), stat[val1.String()])
 	assert.Equal(t, int64(14), stat[val2.String()])
 }
+
+func TestCleanUpFinishedRuns(t *testing.T) {
+	s, err := store.NewStore(nil, 1, tmdb.NewMemDB(), tmdb.NewMemDB())
+	assert.NoError(t, err)
+	assert.NotNil(t, s)
+
+	// test case for lazinessWindow > hibernateThreshold
+	mr := MissRuns{
+		store:              s,
+		hibernateThreshold: 10,
+		hibernatePeriod:    100,
+		lazinessWindow:     12,
+	}
+	val1 := makeValAddr("val1")
+	val2 := makeValAddr("val2")
+	mval00 := []crypto.Address{}
+	mval10 := []crypto.Address{val1}
+	mval12 := []crypto.Address{val1, val2}
+	// mval02 := []crypto.Address{val2}
+
+	mr.UpdateMissRuns(10, mval00)
+
+	stat := mr.GetMissStat(10, 10)
+	assert.Empty(t, stat)
+
+	// val1: hibernating
+	// val2: non-hibernating
+	mr.UpdateMissRuns(11, mval10)
+	mr.UpdateMissRuns(12, mval12)
+	mr.UpdateMissRuns(13, mval12)
+	mr.UpdateMissRuns(14, mval12)
+	mr.UpdateMissRuns(15, mval12)
+	mr.UpdateMissRuns(16, mval12)
+	mr.UpdateMissRuns(17, mval12)
+	mr.UpdateMissRuns(18, mval12)
+	mr.UpdateMissRuns(19, mval12)
+	mr.UpdateMissRuns(20, mval12) // val1 hibernates
+
+	stat = mr.GetMissStat(10, 20)
+	assert.Equal(t, int64(10), stat[val1.String()])
+	assert.Equal(t, int64(9), stat[val2.String()])
+
+	mr.UpdateMissRuns(21, mval00)
+	stat = mr.GetMissStat(10, 21)
+	_, exist := stat[val1.String()]
+	assert.True(t, exist)
+	_, exist = stat[val2.String()]
+	assert.True(t, exist)
+
+	mr.UpdateMissRuns(22, mval00)
+	stat = mr.GetMissStat(10, 22)
+	_, exist = stat[val1.String()]
+	assert.True(t, exist)
+	_, exist = stat[val2.String()]
+	assert.True(t, exist)
+
+	mr.UpdateMissRuns(23, mval00) // val1's missRun gets removed
+	stat = mr.GetMissStat(10, 23)
+	_, exist = stat[val1.String()]
+	assert.False(t, exist)
```

Comment:
No! laziness window (length 12) and val1's missRun are still overlapping. This missRun should be removed at height 32: tail of the run = 20, head of the window = 32 - 12 + 1 = 21 ==> now the window and the runs do not overlap. Keep in mind that clean-up code should not affect the result of `GetMissStat()`.